### DISCUSSION TITLE
How to get total file size storage utilized?

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#2806] [Q&A]
I would like to check user wise files storage utilized. How to get that data?

**Comment by ganigeorgiev** [comment]
There is no builtin functionality at the moment for this and in general it could be difficult to calculate it dynamically, especially if you store your files in a S3 storage.

Additionally, by _"user files"_ do you mean files uploaded to a `users` collection or also files from other collections related to a single user?

In general all uploaded files have path/prefix that follows the format `{collectionId}/{recordId}/{filename}` (_thumbs are stored in a `thumbs_` directory next to the filename_).

You can get the total file size on collection or record level by using something like:

```go
collection, err := app.Dao().FindCollectionByNameOrId("example")
if err != nil {
	return err
}

// "prefix" is the directory path to the file
//
// individual records also have BaseFilesPath() method that you can use
// if you want to get the total files size for all uploaded record files
prefix := collection.BaseFilesPath()

fsys, err := app.NewFilesystem()
if err != nil {
	return err
}
defer fsys.Close()

objs, err := fsys.List(prefix)
if err != nil {
	return err
}

var totalSize int64

for _, obj := range objs {
	totalSize += obj.Size
}
```

But keep in mind that the above may not be very efficient.

Alternatively, to avoid calculating it every time, one workaround could be to attach to the before create/update/delete request and manually add/subtract some `number` field when a new file is uploaded/deleted.

**Reply by iamdineshkumar** [reply]
Hope adding in hooks to calculate and store size manually will be better solution but how to get file size in case file is uploaded? My use case is I am allowing users to upload documents supporting to claims but I wish to know user wise storage utilization. So in my claims collection there is a field for attachment which also allow multiple files to upload.

**Reply by ganigeorgiev** [reply]
You can adjust the above example and use the record prefix to handle both uploaded and deleted files:

```go
// adjust based on your collection requirements
app.OnRecordAfterCreateRequest("claims").Add(func(e *core.RecordCreateEvent) error {
	fsys, err := app.NewFilesystem()
	if err != nil {
		return err
	}
	defer fsys.Close()

	objs, err := fsys.List(e.Record.BaseFilesPath())
	if err != nil {
		return err
	}

	var totalSize int64

	for _, obj := range objs {
		totalSize += obj.Size
	}

	log.Println(totalSize)

	return nil
})
```

Alternatively, you can also access the uploaded files size using the `UploadedFiles` field.
```go
app.OnRecordAfterCreateRequest("claims").Add(func(e *core.RecordCreateEvent) error {
	var totalSize int64

	for _, files := range e.UploadedFiles {
		for _, f := range files {
			totalSize += f.Size
		}
	}

	return nil
})
```
The drawback of the above is that we don't store information of the deleted files and you'll have to somehow fetch them in a before hook and store the size request context (`e.HttpContext.Set()/Get()`) and then perform your calculations in the after hook.

**Reply by iamdineshkumar** [reply]
Whether same code can be used incase files are stored in s3?

**Reply by ganigeorgiev** [reply]
Yes, the above uses the filesystem abstraction and it should work with both S3 or the local filesystem.

**Reply by ganigeorgiev** [reply]
_One minor detail I forgot to mention in case you decide to store the `totalSize` in a `number` field is that it behaves similarly to the JS number, meaning that it can store up to 2^53 int, so in this case you can either use a regular `text` field and store it as string (you still can get a number using `record.GetInt()`) or convert the bytes to a bigger unit (eg. MB) to ensure that there is no precision loss._

**Reply by iamdineshkumar** [reply]
I have decided to store attachmentSize as a field in same record and thought to create a view collection to arrive total value by grouping users. So as you said if I storing in text field whether sum values by grouping userId will workout?

**Reply by ganigeorgiev** [reply]
Yes, the SQLite builtin `sum` should work with text values (note that when using an expression inside a view you'll have to provide a field name as alias for the column, eg. `sum(size) as total`).
